src/ray/raylet/node_manager.cc

```diff
@@ -1584,48 +1584,53 @@ void NodeManager::HandleRequestWorkerLease(const rpc::RequestWorkerLeaseRequest
-    // reject it.
-    ResourceSet normal_task_resources = cluster_task_manager_->CalcNormalTaskResources();
-    RAY_LOG(DEBUG) << "Reject leasing as the raylet has no enough resources."
-                   << " actor_id = " << actor_id
-                   << ", normal_task_resources = " << normal_task_resources.ToString()
-                   << ", local_resoruce_view = "
-                   << cluster_resource_scheduler_->GetLocalResourceViewString();
-    auto resources_data = reply->mutable_resources_data();
-    resources_data->set_node_id(self_node_id_.Binary());
-    resources_data->set_resources_normal_task_changed(true);
-    auto &normal_task_map = *(resources_data->mutable_resources_normal_task());
-    normal_task_map = {normal_task_resources.GetResourceMap().begin(),
-                       normal_task_resources.GetResourceMap().end()};
-    resources_data->set_resources_normal_task_timestamp(absl::GetCurrentTimeNanos());
+    if (is_actor_creation_task) {
+      // If resources are not enough due to normal tasks' preemption, return a rejection
+      // with normal task resource usages.
+      ResourceSet normal_task_resources =
+          cluster_task_manager_->CalcNormalTaskResources();
+      RAY_LOG(DEBUG) << "Reject leasing as the raylet has no enough resources."
+                     << " actor_id = " << actor_id
+                     << ", normal_task_resources = " << normal_task_resources.ToString()
+                     << ", local_resoruce_view = "
+                     << cluster_resource_scheduler_->GetLocalResourceViewString();
+      auto resources_data = reply->mutable_resources_data();
+      resources_data->set_node_id(self_node_id_.Binary());
+      resources_data->set_resources_normal_task_changed(true);
+      auto &normal_task_map = *(resources_data->mutable_resources_normal_task());
+      normal_task_map = {normal_task_resources.GetResourceMap().begin(),
+                         normal_task_resources.GetResourceMap().end()};
+      resources_data->set_resources_normal_task_timestamp(absl::GetCurrentTimeNanos());
+    }
 
     send_reply_callback(Status::OK(), /*success=*/nullptr, /*failure=*/nullptr);
   };
 
-  // If resources are not enough due to normal tasks' preemption, return a rejection with
-  // normal task resource usages.
-  if (!cluster_task_manager_->IsLocallySchedulable(task)) {
+  // GCS only requests actor lease when it thinks that the node has available resources.
+  // If it turns out to be not the case, we should reject immediately
+  // so GCS can update its resource view of this raylet.
+  if (is_actor_creation_task && !cluster_task_manager_->IsLocallySchedulable(task)) {
     reply->set_rejected(true);
     send_reply_callback_wrapper(Status::OK(), /*success=*/nullptr, /*failure=*/nullptr);
     return;
   }
 
-  cluster_task_manager_->QueueAndScheduleTask(task, reply, send_reply_callback_wrapper);
+  cluster_task_manager_->QueueAndScheduleTask(task, true, reply,
```

Comment:
How about avoiding calling QueueAndSchedule for grant_or_reject tasks?

Reply:
We still need to do normal QueueAndSchedule since we still need things like resolving/waiting args.

Reply:
Could probably simplify this further by using the same `send_reply_callback_wrapper` for both grant_or_reject = true/false. Right now, looks like we're calling `send_reply_callback` in 3 different places and it'd be nice to get this down to one for clarity.

Reply:
Good point. Changed to always use `send_reply_callback_wrapper`